Source/Core/Core/PowerPC/SignatureDB/MEGASignatureDB.cpp

```diff
@@ -0,0 +1,162 @@
+    if (GetCode(&sig, &iss) && GetName(&sig, &iss) && GetRefs(&sig, &iss))
+    {
+      m_signatures.push_back(sig);
+    }
+    else
+    {
+      WARN_LOG(OSHLE, "MEGA database failed to parse line %zu", i);
+    }
+  }
+  return true;
+}
+
+void MEGASignatureDB::Apply(PPCSymbolDB* symbol_db) const
+{
+  for (auto& it : symbol_db->AccessSymbols())
+  {
+    u32 hash = it.first;
+    auto& symbol = it.second;
+    for (const auto& sig : m_signatures)
+    {
+      if (Compare(symbol.address, symbol.size, sig))
+      {
+        symbol.name = sig.name;
+        INFO_LOG(OSHLE, "Found %s at %08x (size: %08x)!", sig.name.c_str(), symbol.address,
+                 symbol.size);
+        break;
+      }
+    }
+  }
+  symbol_db->Index();
+}
+
+void MEGASignatureDB::List() const
+{
+  for (const auto& entry : m_signatures)
+  {
+    DEBUG_LOG(OSHLE, "%s : %zu bytes", entry.name.c_str(), entry.code.size() * sizeof(u32));
+  }
+  INFO_LOG(OSHLE, "%zu functions known in current MEGA database.", m_signatures.size());
+}
+
+bool MEGASignatureDB::GetCode(MEGASignature* sig, std::istringstream* iss) const
+{
+  std::string code;
+  if ((*iss >> code) && (code.length() % 8) == 0)
+  {
+    for (size_t i = 0; i < code.length(); i += 8)
+    {
+      std::string instruction = code.substr(i, 8);
+      u32 num = static_cast<u32>(strtoul(instruction.c_str(), nullptr, 16));
+      if (num != 0 || instruction == "........")
+      {
+        sig->code.emplace_back(num);
+      }
+      else
+      {
+        WARN_LOG(OSHLE, "MEGA database failed to parse code");
+        return false;
+      }
+    }
+    return true;
+  }
+  return false;
+}
+
+bool MEGASignatureDB::GetFunctionName(std::istringstream* iss, std::string* name) const
+{
+  std::string buffer;
+
+  std::getline(*iss, buffer);
+  size_t next = buffer.find(" ^");
+  *name = StripSpaces(buffer.substr(0, next));
+
+  if (name->length() == 0)
+    return false;
+
+  if (next == std::string::npos)
+    next = buffer.length();
+  iss->str(buffer.substr(next));
+  iss->clear();
+  return true;
+}
+
+bool MEGASignatureDB::GetName(MEGASignature* sig, std::istringstream* iss) const
+{
+  std::string unknown;
+  return (*iss >> unknown) && GetFunctionName(iss, &sig->name);
+}
+
+bool MEGASignatureDB::GetRefs(MEGASignature* sig, std::istringstream* iss) const
+{
+  std::string num, ref;
+  u32 ref_count = 1;
+  while (*iss && (*iss >> num) && num.length() != 0)
+  {
+    num = num.substr(1);
+    const char* ptr = num.c_str();
+    char* endptr;
+    u64 offset = strtoul(ptr, &endptr, 16);
+
+    if (ptr == endptr || offset > UINT32_MAX)
+    {
+      WARN_LOG(OSHLE, "MEGA database failed to parse reference %u offset", ref_count);
+      return false;
+    }
+
+    if (!GetFunctionName(iss, &ref))
+    {
+      WARN_LOG(OSHLE, "MEGA database failed to parse reference %u name", ref_count);
+      return false;
+    }
+    sig->refs.emplace_back(static_cast<u32>(offset), ref);
+
+    ref_count += 1;
+    num = "";
```

Comment:
A [better](https://godbolt.org/g/ACK3bp) way of doing this would be to use the provided `clear()` function for strings.

```cpp
num.clear();`
ref.clear();
```